thrd_rcu/thrd_rcu.h

```diff
@@ -0,0 +1,215 @@
+#ifndef __THRD_RCU_H__
+#define __THRD_RCU_H__
+
+/* lock primitives from pthread and compiler primitives */
+
+#include <pthread.h>
+
+typedef pthread_mutex_t spinlock_t;
+
+#define SPINLOCK_INIT PTHREAD_MUTEX_INITIALIZER
+
+static __inline__ void spin_lock(spinlock_t *sp)
+{
+    int ret;
+
+    ret = pthread_mutex_lock(sp);
+    if (ret != 0) {
+        fprintf(stderr, "spin_lock:pthread_mutex_lock %d\n", ret);
+        abort();
+    }
+}
+
+static __inline__ void spin_unlock(spinlock_t *sp)
+{
+    int ret;
+
+    ret = pthread_mutex_unlock(sp);
+    if (ret != 0) {
+        fprintf(stderr, "spin_unlock:pthread_mutex_unlock %d\n", ret);
+        abort();
+    }
+}
+
+#define current_tid() (unsigned int) pthread_self()
+
+#define ACCESS_ONCE(x) (*(volatile __typeof__(x) *) &(x))
+#define READ_ONCE(x)                         \
+    ({                                       \
+        __typeof__(x) ___x = ACCESS_ONCE(x); \
+        ___x;                                \
+    })
+#define WRITE_ONCE(x, val)      \
+    do {                        \
+        ACCESS_ONCE(x) = (val); \
+    } while (0)
+#define barrier() __asm__ __volatile__("" : : : "memory")
+#define __allow_unused __attribute__((unused))
+#define smp_mb() __asm__ __volatile__("mfence" : : : "memory")
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#ifdef __CHECKER__
+#define __rcu __attribute__((noderef, address_space(__rcu)))
+#define rcu_check_sparse(p, space) ((void) (((typeof(*p) space *) p) == p))
+#define __force __attribute__((force))
+#define rcu_uncheck(p) ((__typeof__(*p) __force *) p)
+#define rcu_check(p) ((__typeof__(*p) __force __rcu *) p)
+#else
+#define __rcu
+#define rcu_check_sparse(p, space)
+#define __force
+#define rcu_uncheck(p) p
+#define rcu_check(p) p
+#endif /* __CHECKER__ */
+
+/* Avoid false sharing */
+#define __rcu_aligned __attribute__((aligned(128)))
+
+struct rcu_node {
+    unsigned int tid;
+    int rcu_nesting[2];
+    struct rcu_node *next;
+} __rcu_aligned;
+
+struct rcu_data {
+    unsigned int nr_thread;
+    struct rcu_node *head;
+    unsigned int rcu_thrd_nesting_idx;
+    spinlock_t sp;
+};
+
+/* Easy to use */
+#define __rcu_thrd_idx rcu_data.rcu_thrd_nesting_idx
+#define __rcu_thrd_nesting(ptr) \
+    ptr->rcu_nesting[READ_ONCE(__rcu_thrd_idx) & 0x01]
+#define rcu_thrd_nesting __rcu_thrd_nesting(__rcu_per_thrd_ptr)
+
+static struct rcu_data rcu_data = {.nr_thread = 0,
+                                   .head = NULL,
+                                   .rcu_thrd_nesting_idx = 0,
+                                   .sp = SPINLOCK_INIT};
```

Comment:
Add a comma (`,`) after `SPINLOCK_INIT`, so that clang-format can indent much better.